### ISSUE TITLE
CppFlow with pretrained models from 'Tensorflow model zoo'

### ISSUE THREAD

**Issue by Genyz** [serizba/cppflow#1]
Hello,
Interresting and usefull job. Thanks for sharing it !

I've tried to load with CppFlow a Tensorflow pretrained 'ssd_inception_v2_coco' model (for example, but I tried several other models) found here :
https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/detection_model_zoo.md

I wanted to do, in C++, something like what you can find here in Python :
https://github.com/opencv/opencv/blob/master/samples/dnn/mobilenet_ssd_accuracy.py
(or in the attached file)
[my_ssd_accuracy.py.txt](https://github.com/serizba/cppflow/files/3233419/my_ssd_accuracy.py.txt)

So I tried with CppFlow a code like this :
	Model model("ssd_inception_v2_coco_2018_01_28/frozen_inference_graph.pb");

That works fine.

Then :

	auto outNames1 = new Tensor(model, "num_detections");
	auto outNames2 = new Tensor(model, "detection_scores");
	auto outNames3 = new Tensor(model, "detection_boxes");
	auto outNames4 = new Tensor(model, "detection_classes");

	auto inpName = new Tensor(model, "image_tensor");

But I was not able to make it work. Seems that there are errors after such functions as 'TF_GraphGetTensorShape' or 'TF_GraphGetTensorNumDims'.

I'm not a Tensorflow specialist but rather a user. Maybe am I wrong somewhere and it's not possible to load this kind of pretrained model.
Can you confirm it's possible ?

Thanks a lot !

**Comment by serizba** [comment]
The problem is that the code is prepared to work with STD Vectors, plain vectors without shape, and therefore it throws and exception if the dimension is not known. I have added a function to load data saying the exact shape you want to provide, for example, in your case: 

```c++
inpName->set_data(img, {1, 300, 300, 3});
```

Now you can declare inputs or outputs without known shape until you feed data or run the model. I have tried your code and it works perfectly (although I have not tested every case). I have uploaded an example using the  'ssd_inception_v2_coco' to the examples folder (copy of your Python example but in C++). I hope it is useful for you!

Thanks you for sharing this error.

**Comment by Genyz** [comment]
Hi.
I could process tests this morning.
It works fine. Thanks a lot for your investigations !

Just a few remarks :

1 -
I'm working with Microsoft Visual Studio 2017 and Windows 10
To avoid a compilation error, I had to comment line 208 in Tensor.cpp
//template std::vector<bool> Tensor::get_data<bool>();

2 -
My purpose for using C code instead of Python code was to evaluate the improvement in processing speed during inference.
I'm a bit surprised because seems that the processing time is slower using CppFlow (71 msec per image with code running on my Core i5), compared to Python (51 msec per image), using exactly the same .pb files and test images in C and Python.
Maybe would you have a suggestion.

Thanks !

**Comment by serizba** [comment]
Well, the only advantage in using C++ would be in loading and processing the data before feeding it to the network. 

You should take into account that although your program is written in Python, you are only designing the Network architecture. Tthe inference, the real network code will be created by Tensorflow in C++, optimized C and cuda.

When you use the inference from C++ you will be running the same network and the same code as with Python, the only part is different is how you read the data and feed to the network. CppFlow is just a simple wrapper, and therefore, when you set data into a Tensor, all the data is being copied, and then given to the network. The same when you get the data. In Python, everything is already prepared to use Numpy arrays and optimized for this inputs. I think that the translation between std vector to TensorFlow Tensors  is the overhead that makes the difference in execution time.

**Comment by Genyz** [comment]
Thanks for your explanations.
I will continue to follow all your work and github !
